climada/hazard/base.py

```diff
@@ -1393,3 +1442,167 @@ def _read_att_excel(self, file_name, var_names):
+            hazards from haz_src. The default is False.
+        centroids: climada.hazard.Centroids(), optional
+            Centroids instance on which to map the concatenated hazard.
+            Default is None.
+
+        Returns
+        -------
+        None.
+
+        """
+        if append:
+            haz_src = [self] + haz_src
+        self.__dict__ = copy.deepcopy(
+            concatenate_hazard(haz_src, centroids).__dict__
+            )
+
+
+def join_centroids(haz_list):
+    """
+    Find centroids comon to all.
+
+    Centroids of all hazards must either all be rasters with the same
+    resolution, or all be points.
+
+    The centroids of all hazards are combined together. If points,
+    all points are combined. If rasters, a global raster is defined.
+
+    Parameters
+    ----------
+    haz_list : list(climada.hazard.Hazard())
+        List of hazard with centroids to join
+
+    Returns
+    -------
+    centroids : climada.hazard.Centroids()
+        Centroids containing the centroids of all hazards.
+
+    """
+    centroids = copy.deepcopy(haz_list[0].centroids)
+    for haz in haz_list[1:]:
+        if not centroids.equal(haz.centroids):
+            centroids.append(haz.centroids)
+    centroids.remove_duplicate_points()
+    return centroids
+
+def concatenate_hazard(haz_list, centroids=None):
+    """
+    Concatenate events of several hazards of same type.
+
+    Centroids of all hazards must either all be rasters with the same
+    resolution, or all be points.
+
+    By default, the centroids of all hazards are combined together. If points,
+    all points are combined. If rasters, a global raster is defined.
+
+    If centroids is not None, all hazards are defined on these centroids.
+    Centoids of hazards in haz_list not in centroids are mapped onto the
+    nearest point.
+
+    Parameters
+    ----------
+    haz_list: list(climada.hazard.Hazard())
+        Hazard instances of the same hazard type
+    centtroids: climada.hazard.Centroids(), optional
+        Centroids instance on which to map the concatenated hazard.
+        Default is None.
+
+    Returns
+    -------
+    haz_concat: climada.hazard.Hazard()
+        Concatenated hazard.
+    """
+    #Check units consistency among hazards
+    units = {haz.units for haz in haz_list}
+    if len(units) > 1:
+        raise TypeError("The haz_list contains hazards with different"
+                        "units %f. The hazards are incompatible and"
+                        "cannot be concatenated.", units)
+
+    #Define comon centroids
+    if centroids is None:
+        centroids = join_centroids(haz_list)
+
+    haz_concat = Hazard()
+    haz_concat.units = haz_list[0].units
+    haz_concat.centroids = centroids
+
+    #Indices for mapping matrices onto common centroids
+    hazcent_in_cent_idx_list = [
+        u_coord.assign_coordinates(haz.centroids.coord, centroids.coord)
+        for haz in haz_list
+        ]
```

Comment:
It might be useful if this object was somehow made available to the user? If I have some large exposures that I spent ages mapping to centroids, it'd be great to be able to update the with this instead of having to run `assign_coordinates` again.

Reply:
Interesting idea!

Two potential objections:
- it is quite an edge case, and returning this value only for these edge cases is a strong hit to user friendliness imho.
- `u_coord.assign_coordinates` is not thaaat time consuming normally. In particular, if the centroids and the exposures lat/lon match, it is very fast.

What do you think?

Reply:
If you're working with (near-)global hazards and exposures defined on different grids `assign_coordinates` takes forever for me. So I'd looove to avoid running it again when combining hazards.

Having said that, I can't see a really user-friendly way to provide this.

One option would be to add `hazcent_in_cent_idx_list` as an additional, optional input parameter (maybe with a better name), and then split these lines into a separate utility function. Then the user has the choice to calculate these mapping manually if they're required, and then pass them to the function. But in in most cases you just call the function and it's calculated as it is right now.

Counter-argument: it's an edge case and it's easy for me to write my a custom method when I need it.

Reply:
Yes, I was also thinking of a separate util function, but came to the conclusion that it is a) quite easy to write b) not broadly useful. Hence, I kept it this way. 

"If you're working with (near-)global hazards and exposures defined on different grids assign_coordinates takes forever for me. So I'd looove to avoid running it again when combining hazards."

But in this case what would you want to concatenate which has different centroids?

Reply:
OK let's leave it as it is then.

(And to answer your question, I don't have a convincing use case yet, beyond maybe combining very high resolution datasets from separate basins)